### ISSUE TITLE
[GR-60494] Add basic support for 32bit JVM with no optimizations

### ISSUE THREAD

**Issue by csterg** [oracle/graal#10294]
Hello,
I am looking to see if there is any way to run graaljs on a stock 32 bit JDK but I can find none. I understand that the architecture is such that several optimizations are hard (or impossible) to do on 32bit, yet there is a real void for applications that need to run on 32bit architectures (yes there are cases that using a 64bit JDK is not an option, e.g. loading native legacy 32bit dll on windows).

A 'plain Java'mode that requires nothing of the underlying JVM in terms of JIT would help move away from Rhino/Nashhorn and embrace new ECMA language features, even at a degraded runtime speed.

Are there plans to do so? 
Is there any way to 'hack' the runtime in a way that will run on 32bit, even by sacrificing optimizations?

Greatly appreciate your help and insights (badly stuck here)
Thanks

**Comment by csterg** [comment]
Following up to my question: stepping blindly through the code, I found that setting the system option:  `-Dtruffle.UseFallbackRuntime=true` (found it in class `Truffle::createRuntime` following the debugger)
avoid the x64 specific startup that loads the 64bit native dll and allows graaljs to run. 

I have absolutely no idea if issues will appear down the line, but is this an alternative to run graaljs in x86 JVM? If yes this is AMAZING and a way to continue having a modern JS engine in latest 32bit stock runtimes.
Thanks

**Comment by selhagani** [comment]
Hi @csterg,

Thank you for reaching out to us about this. We'll take a close look at this and I'll make sure to keep you updated.

**Comment by chumer** [comment]
> Are there plans to do so?

Given that there are plans to remove 32 bit versions all together (https://openjdk.org/jeps/501) I don't think it makes a lot of sense to invest time into being fully 32 bit compliant. Please direct your feedback on the 32 bit deprecation at the Java team.

That being said, if it works with the fallback runtime, you should be good for now. Note that it is likely given the upcoming removals of the the sun.misc.Unsafe APIs that we will need to load the native Truffle attach library even for the fallback runtime with JDK 25, but we are not 100% certain yet. Besides the fallback runtime property you can also exclude the truffle-runtime and truffle-enterprise artefacts like this:
https://www.graalvm.org/latest/reference-manual/embed-languages/#switching-to-the-fallback-engine

Even if we need to load a native library, it would be theoretically possible to ship a 32 bit version of that library in addition to the 64 bit version. But I don't think I can justify the effort for this due to the deprecation. So keep in mind that support for running like this might eventually stop. I will try to mention this in the changelog (https://github.com/oracle/graal/blob/master/sdk/CHANGELOG.md) when the time comes.

Long story short, using the fallback runtime on JDKs prior 25 can be useful to migrate away from Rhino. I don't think you will run into trouble with the fallback runtime, but also keep in mind that we are not testing on that platform, so you are on your own.
